Implement support for requesting light/dark app appearance

vadz opened the pull request:
We've had several requests for this, so I've finally implemented it. This is relatively straightforward, but I had to refactor wxGTK code quite a bit to allow reusing parts of it that I needed.

Any comments about API or the implementation (especially under Mac, where, as usual, I don't really know what I'm doing) would be welcome, TIA!

vadz committed:
Move UpdatePreferDark() function into anonymous namespace

No real changes.

vadz committed:
Move wxSystemSettingsModule::m_proxy to gs_proxyPortalSettings

Use an actually global variable for a variable which was effectively
global before but was only accessible via wxSystemSettingsModule in
order to make it possible to reuse it from outside of it in the future.

No real changes yet.

vadz committed:
Return early from UpdatePreferDark() if there is nothing to do

No real changes.

This commit is best viewed ignoring whitespace-only changes.

vadz committed:
Add tracing to wxGTK dark mode handling code

This can be useful to follow what's being done when things don't work as
expected and shouldn't add any noticeable overhead.

vadz committed:
Add wxApp::SetAppearance() to explicitly set dark or light mode

Allow the application to override the default system mode and,
incidentally, also to provide a more portable way of enabling dark mode
under MSW.

Show how to use the new function in the drawing sample.

Closes #19185.

Closes #23867.

csomor reviewed (commented):
Thanks, the mac part looks ok to me (although that API has been deprecated already in macOS 11 ... as usual short lived ...)

paulcor commented on src/gtk/settings.cpp:
```diff
@@ -183,40 +188,132 @@ static void notify_gtk_font_name(GObject*, GParamSpec*, void*)
+               preferDark ? "on" : "off");
+
+    g_object_set(settings,
+        "gtk-application-prefer-dark-theme", preferDark, nullptr);
```
You can't change the theme this way if `GTK_THEME` is set, as it overrides `GtkSettings`, so `UpdatePreferDarkWithoutRefresh()` should probably not report success in that case.

vadz replied:
Thanks, indeed. Now I wonder if we should have different `CannotChangeBecauseOverriddenByTheme` and `CannotChangeWithoutRestart` return codes (except preferably shorter), as the application would probably want to handle them differently.

paulcor commented on src/gtk/app.cpp:
```diff
@@ -347,6 +351,36 @@ bool wxApp::SetNativeTheme(const wxString& theme)
+wxApp::AppearanceResult wxApp::SetAppearance(Appearance appearance)
+{
+#ifdef __WXGTK3__
+    gboolean preferDark = FALSE;
```
As far as I know, it's perfectly fine to assign `true`/`false` to an `int`, so I don't see any reason not to prefer those over the old `TRUE`/`FALSE`.

vadz replied:
IMO this is more clear: even if C++ allows you to assign just about [anything to anything else](https://fosstodon.org/@VZ/112235960416690642), using `false` with `bool` variables and `FALSE` with `int` ones is IMO tidier and I've always done it like this in wxMSW (which has its own `BOOL` which is actually `int`) and would rather keep doing it in wxGTK too (where there are already more than 330 occurrences of these constants) unless you absolutely hate it.

paulcor commented on src/gtk/settings.cpp:
```diff
@@ -183,40 +188,132 @@ static void notify_gtk_font_name(GObject*, GParamSpec*, void*)
+
+bool GetPreferDarkFromColorScheme(gboolean* preferDark)
+{
+    GVariant* ret = g_dbus_proxy_call_sync(gs_proxyPortalSettings, "Read",
```
It seems possible for `gs_proxyPortalSettings` to be null here.

vadz replied:
Fixed, thanks.

paulcor commented on src/gtk/settings.cpp:
```diff
@@ -183,40 +188,132 @@ static void notify_gtk_font_name(GObject*, GParamSpec*, void*)
+
+// This function fills in the provided pointer (which must be non-null) and
+// returns GtkSettings object which is also used to indicate a failure.
+GtkSettings* GetPreferDark(gboolean* preferDarkPrev)
```
If `preferDarkPrev` must be non-null, it seems like it should be a reference.

vadz replied:
We rarely use output parameters but when we do, we typically use pointers to them in order to clearly indicate that they're been modified in the caller. As I find this more readable I'd like, again, to keep it like this unless you object violently.

PBfordev commented on interface/wx/app.h:
```diff
@@ -965,6 +965,64 @@ class wxApp : public wxAppConsole
+        Under GTK and macOS applications use the system default appearance by
+        default, and so it is only useful to call this function with either
+        Appearance::Light or Appearance::Dark parameters if you need to
+        override the default system appearance. The effect of calling this
```
I wonder if we should note the existence of `wxSystemAppearance` here and analogically, reference this method in `wxSystemAppearance` docs.
```suggestion
        override the default system appearance (see wxSystemAppearance). The effect of calling this
```

The wxWidgets API is very extensive and sometimes it may be hard to find the correct method/class that does what is wanted...

vadz replied:
Thanks, I've added xrefs in both directions now.

vadz commented:
I've addressed the non-style review comments now, thanks!

If there are no other ones, I'll merge this soon.

paulcor commented:
I think there may be another problem with this, at least for GTK. If the user changes the global DE light/dark setting after `SetAppearance()` is called, the `SetAppearance()` setting will be overridden.

vadz committed:
Add wxGtkVariant helper and start using it in GTK settings code

No real changes, just get rid of a bunch of g_variant_unref() calls.

This wrapper could also be made more type-safe in the future, although
this is not done yet.

vadz committed:
Return early from proxy_g_signal() too

Still no real changes, just return early now that we use wxGtkVariant
and don't have to free it manually before exiting the function.

vadz commented:
> I think there may be another problem with this, at least for GTK. If the user changes the global DE light/dark setting after `SetAppearance()` is called, the `SetAppearance()` setting will be overridden.

Thanks, this was indeed a bug and I've fixed it now but I had to change the code a lot in order to do it and so I've also squashed some commits to get rid of some refactorings that turned out to not be really useful any more.

I also had to introduce RAII `wxGtkVariant` helper here, which I wanted to postpone until later, but without it the code became more difficult to write, so I've finally done it here.

Thanks again for reviewing this!
